### PR TITLE
[dashboards] add support for storage parameter in widget queries

### LINKED ISSUE #1543: [Feature request] Support for `storage` attribute on dashboard log queries
The Datadog Dashboard and querying UI allow us to choose either `Indexes` or `Online Archives` storage options.
image

However, that is not supported yet on the terraform API.  At least as far as I can see by checking the dashboard API guideline and repository documentation; https://registry.terraform.io/providers/DataDog/datadog/latest/docs/resources/dashboard

There is no mention of `storage`, and naturally, the terraform API complains about attempts of its usage.

### Terraform Version

Terraform v1.2.4
registry.terraform.io/datadog/datadog v3.14.0

### Affected Resource(s)
-   datadog_dashboard

### Terraform Configuration Files

```hcl
#...
event_query {
  data_source = "logs"
  storage = "online_archives"
  indexes = [
    "*",
  ]
  name = "query1"

  compute {
    aggregation = "count"
    interval    = 0
  }

  search {
    query = "..."
  }
}
# ...
```

### Expected Behavior

The TF API should have accepted the `storage` option and highlighted it in the changes.

### Actual Behavior
The TF API complains about the unsupported storage argument.
```
╷
│ Error: Unsupported argument
│
│   on modules/datadog/dashboard.tf line 681, in resource "datadog_dashboard" "edge":
│  681:             storage = "online_archives"
│
│ An argument named "storage" is not expected here.

```

### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:

1. Create a `datadog_dashboard`.
2. Create a `query`.
3. Create an `event_query`, use `data_source = logs` fill required fields.
4. Try adding `storage = "online_archives"`.
5. Try `terraform plan`

### PR DESCRIPTION
**Pull request by nkzou**
Resolves #1543. Adds support for the storage parameter in dashboard widgets that use it.

### CODE REVIEW

File: docs/resources/dashboard.md [hunk 1]
```diff
@@ -1103,6 +1103,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--change_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--change_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:1106]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 2]
```diff
@@ -1939,6 +1940,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--geomap_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--geomap_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:1943]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 3]
```diff
@@ -2477,6 +2479,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--group_definition--widget--change_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--group_definition--widget--change_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:2482]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 4]
```diff
@@ -3313,6 +3316,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--group_definition--widget--geomap_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--group_definition--widget--geomap_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:3319]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 5]
```diff
@@ -4498,6 +4502,7 @@ Optional:
 
 - `indexes` (List of String) List of indexes.
 - `query_string` (String) Widget query.
+- `storage` (String) Storage location. Private Beta feature.
```
Comment (alai97) [inline docs/resources/dashboard.md:4505]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 6]
```diff
@@ -8327,6 +8337,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--group_definition--widget--toplist_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--group_definition--widget--toplist_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:8340]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 7]
```diff
@@ -10284,6 +10297,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--query_table_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--query_table_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:10300]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 8]
```diff
@@ -10875,6 +10889,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--query_value_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--query_value_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:10892]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 9]
```diff
@@ -12398,6 +12414,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--sunburst_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--sunburst_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:12417]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 10]
```diff
@@ -13077,6 +13094,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--timeseries_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--timeseries_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:13097]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 11]
```diff
@@ -13697,6 +13715,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--toplist_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--toplist_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:13718]:
```suggestion
- `storage` (String) Storage location (private beta).
```

File: docs/resources/dashboard.md [hunk 12]
```diff
@@ -14109,6 +14128,7 @@ Optional:
 - `group_by` (Block List) Group by options. (see [below for nested schema](#nestedblock--widget--treemap_definition--request--query--event_query--group_by))
 - `indexes` (List of String) An array of index names to query in the stream.
 - `search` (Block List, Max: 1) The search options. (see [below for nested schema](#nestedblock--widget--treemap_definition--request--query--event_query--search))
+- `storage` (String) Storage location. Private Beta feature
```
Comment (alai97) [inline docs/resources/dashboard.md:14131]:
```suggestion
- `storage` (String) Storage location (private beta).
```

**Review by alai97 — APPROVED** [review]
Looks good after this general formatting nit!

**Review by DrkSephy — APPROVED** [review]

**Review by skarimo — COMMENTED** [review]
Hmm looks like the definitions of the requests didn't change so we probably don't need all of these cassette changes. However, could you add a test case for 'storage' option for FormulaAndFunctions query, thanks

**Comment by zmotu** [comment]
Can't wait for this next release to be published so we can start making use of this.

**Comment by zmotu** [comment]
@skarimo would it be possible to expedite a new release to start taking advantage of this new feature.

**Comment by skarimo** [comment]
@zmotu We will try and release early next week!

**Pull request merged**